RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+| `/<name>/template.yml`             | YAML file describing the template |
+| `/<name>/template_schema.json` (1) | JSON Schema file describing the valid settings for this template |
+| `/<name>/files/**/*`               | Templates to process for new module, convert and update |
+| `/<name>/files_init/**/*`      (1) | Templates to process when the module does not yet exist |
+| `/<name>/*`                  (1,2) | Additional files used _for_ templating, for example configuration information |
+
+(1) Optional
+
+(2) Ruby files that will be evaluated for the ERB templating process **must** be specified in the `template.yml` file. This is to help (somewhat) safeguard arbitrary Ruby being evaluated. Although it can be argued that ERB templates are also arbitrary code.
+
+> TODO: Does files_init EVEN make sense?!??!
+> Actually it does it in an `always_apply` template
+
+> TODO: What about object_templates?
+
+Note - `/object_templates/*.erb` will still be consumed by the V2 Templating Engine.
+
+> TODO: What about `/rubocop/` ?
+
+### `<name>/template.yml`
+
+| YAML Element | Description |
+| ------------ | ----------- |
+| name (required) | A short descriptive name of the template |
+| description | A long (even multiline) description of the template. Default is the name of the template |
+| type | What type of item this is a template for. Currently only `module` is supported, but in time `control_repo` and others may be added. Default is `module` |
+| evaluate_ruby_files | An array of files to [evaluate](https://ruby-doc.org/core-2.6.4/Kernel.html#method-i-eval) prior to rendering ERB templates. This is used to inject complex logic into the templating system without the troubles of having complex logic in ERB.  For example complex branching based on settings. Default is empty array |
+| default_settings | A YAML hash of the default settings for the template. Default is empty hash |
+| always_apply | Always apply this template. If not specified in the template list, it will be applied last. Default is `false` |
+| tags | An array of string tags for this template. Always includes the directory name of the module. Default is an empty array |
+
+### `<name>/template_schema.json`
+
+### Example templates
+
+#### A template for a default module file
+
+The `.pdkignore` file will be used in all PDK compatible modules therefore it should always be applied.
+
+File layout
+
+``` text
+pdkignore/
+  +- template.yml
+  +- template_schema.json
+  +- files/
+       +- .pdkignore.erb
+```
+
+
+`template.yml`
+
+``` yaml
+---
+name: "PDK Ignore"
+description: >-
+  The PDK Ignore template manages the .pdkignore file for Puppet Modules
+type: module
+always_apply: true
+tags: ignore default
+default_settings:
+  required:
+    - '/appveyor.yml'
```

Comment:
this reminds me of a missing requirement: being able to share config data between templates. this same list should also go into the `.gitignore` file.

Reply:
I think there should be a higher level of shared config settings, but I'm not sure about templates being able to export variables or directly reference settings from other templates. I think "hoisting" things up to the shared scope should be a thoughtful decision.

It seems easier to subsequently hoist something up to the shared scope and update a template to default to the shared setting than to untangle interdependencies between template components.